// gestures/walk.js
var three = require('three');
var world = require('world');
var physics = require('physics');
var view = require('view');

var ball;
var avatar;
var lastX = 0;
var walkGesture = {
	touches: 1,
	name: 'walk',

	onStart: onStart,
	onMove: onMove,
	onEnd: onEnd,
}

function onStart(touches) {
	avatar = physics.getAvatar();
	console.log('external walk start', touches.length);
	var x = touches[0].clientX;
	lastX = x;
	var y = view.height - touches[0].clientY;
	if(!ball) {
		ball = new three.Mesh(new three.SphereGeometry(50, 32, 16));
	}
	if(!ball.parent) {
		world.scene.add(ball);
	}
	ball.position.set(view.camera.left + x, view.camera.bottom + y, 0);
}

function onMove(touches) {
	console.log('external walk move', touches.length);
	var x = touches[0].clientX;
	var y = view.height - touches[0].clientY;
	var deltaX = x - lastX;
	if(Math.abs(deltaX) > 20) walk(deltaX);
	else walk(0);
	lastX = x;
	ball.position.set(view.camera.left + x, view.camera.bottom + y, 0);
}

function onEnd(touches) {
	console.log('external walk end', touches.length);
	var x = touches[0].clientX;
	var y = view.height - touches[0].clientY;
	if(ball.parent) {
		world.scene.remove(ball);
	}
	ball.position.set(view.camera.left + x, view.camera.bottom + y, 0);
}

function walk(delta) {
	avatar.walk(delta);
}
module.exports = walkGesture;
