// scenery/TestPlane.js
var three = require('three');
var createPlane = require('geometry/create-plane');

var vertexShader = [
	"varying float d;",
	"void main() {",
	"	gl_Position = projectionMatrix * modelViewMatrix * vec4(position, 1.0);",
	"	d = gl_Position.z * gl_Position.z * gl_Position.z;",
	"}"
].join("\n");

var fragmentShader = [
	"varying float d;",
	"uniform vec3 color;",
	"uniform vec3 color2;",
	"void main() {",
	"	gl_FragColor = vec4(mix(color, color2, d * 4.0), 1.0);",
	"}"
].join("\n");


var geometry;

function getGeometry() {
	if(!geometry) {
		geometry = createPlane();
	}
	return geometry;
}

var material;
function getMaterial() {
	if(!material) {
		var uniforms = {
			color: { type: 'c', value: new THREE.Color(0.5, 0.6, 0.5) },
			color2: { type: 'c', value: new THREE.Color(0.8, 0.85, 0.9) }
		}
		material = new three.ShaderMaterial({
			uniforms: uniforms,
			fragmentShader: fragmentShader,
			vertexShader: vertexShader
		});
	}
	return material;
}

function TestPlane() {
	three.Mesh.call(this, getGeometry(), getMaterial());
}

TestPlane.prototype = Object.create(three.Mesh.prototype);

module.exports = TestPlane;
