// view/index.js
var resize = require('input-resize');
var three = require('three');
var screenfull = require('screenfull');
var EventUtil = require('browser-event-adder');
var urlParam = require('urlparam');
var colors = require('colors');
THREE = three;

function noop() {}
var initd = false;

var renderer;
var screenAspectRatio = 1;
var resizeHandlers = [];

var preRender = noop;
var renderPasses = [];
var onEnterFrame = noop;

function onResize(w, h) {
	renderer.setSize(w, h);
	screenAspectRatio = h / w;
	exports.aspectRatio = screenAspectRatio;
	exports.screenWidth = w;
	exports.screenHeight = h;
	resizeHandlers.forEach(function(handler) {
		handler(w, h);
	});
}

function addResizeHandler(handler) {
	resizeHandlers.push(handler);
	handler(exports.screenWidth, exports.screenHeight);
}

function addRenderPass(renderPass) {
	renderPasses.push(renderPass);
}

function rafTick() {
	onEnterFrame();
	preRender();
	renderPasses.forEach(function(renderPass) {
		renderPass.render(renderer);
	});
	window.requestAnimationFrame(rafTick);
}

function setOnEnterFrame(callback) {
	onEnterFrame = callback;
}

function setPreRender(callback) {
	preRender = callback;
}

function init() {
	if(initd) return;
	initd = true;
	var canvas = require('canvas');

	renderer = new three.WebGLRenderer({
		canvas: canvas.canvas,
	});
	var colorSky = new three.Color(0.8, 0.85, 0.9);
	renderer.setClearColor(colorSky);

	var pixelRatio = window.devicePixelRatio;
	// var pixelRatio = 0.5;
	renderer.setPixelRatio(pixelRatio);

	exports.pixelRatio = pixelRatio;
	resize.onResize.add(onResize);
	resize.bump(onResize);

	window.requestAnimationFrame(rafTick);

	if(!urlParam('windowed')) {
		EventUtil.addEvent(canvas.canvas, 'touchend', requestFullscreen);
	}

	delete exports.init;
	exports.setPreRender = setPreRender;
	exports.setOnEnterFrame = setOnEnterFrame;
	exports.requestFullscreen = requestFullscreen;
	exports.addResizeHandler = addResizeHandler;
	exports.addRenderPass = addRenderPass;
}

function requestFullscreen() {
	try {
		if (screenfull.enabled) {
			screenfull.request();
		}
	} catch(e) {
		console.error(e);
	}
}

var exports = {
	screenWidth: 100,
	screenHeight: 100,
	init: init
};

module.exports = exports;
