// view/index.js
var resize = require('input-resize');
var three = require('three');
var screenfull = require('screenfull');
var EventUtil = require('browser-event-adder');
var world = require('world');
THREE = three;
var Checkerboard = require('threejs-texture-checkerboard');

var initd = false;

var renderer;
var camera;

var testPlane;

function onResize(w, h) {
	renderer.setSize(w, h);
	camera.right = w;
	camera.top = h;
	camera.updateProjectionMatrix();
	module.exports.width = w;
	module.exports.height = h;
}

function rafTick() {
	window.requestAnimationFrame(rafTick);
	renderer.render(world.scene, camera);

}

function init() {
	if(initd) return;
	initd = true;
	var canvas = require('canvas');

	renderer = new three.WebGLRenderer({
		canvas: canvas.canvas
	});

	var pixelRatio = window.devicePixelRatio;
	// var pixelRatio = 0.5;
	renderer.setPixelRatio(pixelRatio);

	module.exports.pixelRatio = pixelRatio;

	camera = new three.OrthographicCamera(0, 100, 100, 0, -100, 100);
	world.scene.add(camera);

	testPlane = new three.Mesh(
		new three.PlaneGeometry(10000, 10000, 1, 1),
		new three.MeshBasicMaterial({
			map: new Checkerboard(undefined, undefined, 100, 100),
		})
	);
	testPlane.position.set(5000, 5000, 0);
	world.scene.add(testPlane);

	resize.onResize.add(onResize);
	resize.bump(onResize);

	renderer.render(world.scene, camera);

	window.requestAnimationFrame(rafTick);

	EventUtil.addEvent(canvas.canvas, 'touchend', requestFullscreen);

	delete module.exports.init;
	module.exports.requestFullscreen = requestFullscreen;
}

function requestFullscreen() {
	if (screenfull.enabled) {
		screenfull.request();
	}
}

module.exports = {
	init: init
};
