// gui/index.js
var three = require('three');
var RenderPass = require('render-utils/RenderPass');
var Heart = require('./HealthHeart');

var scene;
var camera;
var screenAspectRatio = 1;
var heart;

function init() {
	scene = new three.Scene();
	camera = new three.OrthographicCamera(0, 100, 100, 0, -1000, 1000);
	scene.add(camera);

	var ball = new three.Mesh(new three.SphereGeometry(20));
	scene.add(ball);
	ball.position.x = 200;
	ball.position.y = 100;

	heart = new Heart();
	scene.add(heart);
	heart.position.x = 60;
	heart.position.y = 100;

	exports.scene = scene;
	exports.camera = camera;
	exports.getRenderPass = getRenderPass;
	exports.onResize = onResize;
	exports.heart = heart;
}

function getRenderPass() {
	return new RenderPass(scene, camera, false);
}

function onResize(w, h) {
	screenAspectRatio = h / w;
	camera.left = 0;
	camera.top = 0;
	camera.bottom = h;
	camera.right = w;
	camera.updateProjectionMatrix();
	updateLayout();
}

function updateLayout() {

}

var exports = {
	init: init
};

module.exports = exports;
