// game-objects/Enemy.js
var Fighter = require('./Fighter');
var physics = require('physics');
var bits = require('physics/bits');
var defaults = require('lodash.defaults');

var defaultActorParams = {
	name: 'enemy',
	shape: 'circle',
	radius: 25,
	mass: 1,
	friction: 0,
	x: 300,
	categoryBits: bits.enemy,
	maskBits: bits.environment | bits.hero | bits.enemy | bits.heroWeapon,
	fixedRotation: false,
	linearDamping: 10,
	interval: 2000,
	color: 0xbb2222,
	team: 'enemy',
	fistParams: {
		categoryBits: bits.enemyWeapon,
		maskBits: bits.environment | bits.hero,
		power: 0.25,
		speed: 0.25
	}
};

function Enemy(params) {
	params = params || {};
	defaults(params, defaultActorParams);
	Fighter.call(this, params);
}

Enemy.prototype = Object.create(Fighter.prototype);

Enemy.prototype.getClosestInterestingThing = function() {
	return physics.getClosestHeroBody(this.body.GetPosition(), 0.5);
};

var material;
Enemy.prototype.getMaterial = function(color) {
	if(!material) {
		material = new THREE.MeshBasicMaterial({
			color: color,
			transparent: true,
			opacity: 0.5
		});
	}
	return material;
};


module.exports = Enemy;
