// game-objects/SeekerPoint.js
var defaults = require('lodash.defaults');
var physics = require('physics');
var defaultActorParams = {
	shape: 'circle',
	radius: 200,
	mass: 0.0001,
	density: 0.00001,
	x: 500,
	y: 100,
	maskBits: 0,
	categoryBits: 0
};

var tempVec2;
function SeekerPoint(params) {
	params = params || {};
	defaults(params, defaultActorParams);
	this.radius = params.radius;
	this.body = physics.addActor(params);
	if(!params.targetBody) throw new Error('You must provide a target body');
	this.body.SetTransform(params.targetBody.GetPosition(), 0);
	this.joint = physics.addJointSpring({
		bodyA: this.body,
		bodyB: params.targetBody
	});
	this.joint = physics.addJointFriction({
		bodyA: this.body,
		bodyB: params.targetBody,
		maxForce: 0.000012,
	});
	physics.registerPrePhysics(this.fly.bind(this));
	tempVec2 = physics.getTempVec2();
}

SeekerPoint.prototype.fly = function() {
	// tempVec2.set_x(0);
	// tempVec2.set_y(0, 4);
	this.body.ApplyForceToCenter(physics.getTempVec2());
}
module.exports = SeekerPoint;
